plutus-core/executables/uplc/Main.hs

```diff
@@ -259,35 +260,54 @@ runDbg (DbgOptions inp ifmt cekModel) = do
                     -- type is encountered.  This is useful for calibrating the budgeting code
                     Unit    -> PLC.unitCekParameters
         MachineParameters costs runtime = cekparams
+        replSettings = Repl.Settings { Repl.complete = Repl.noCompletion
+                                     , Repl.historyFile = Nothing
+                                     , Repl.autoAddHistory = False
+                                     }
     let ?cekRuntime = runtime
         ?cekEmitter = const $ pure ()
         ?cekBudgetSpender = Cek.CekBudgetSpender $ \_ _ -> pure ()
         ?cekCosts = costs
         ?cekSlippage = D.defaultSlippage
-      in Repl.runInputT (Repl.Settings Repl.noCompletion Nothing False) $
+      in Repl.runInputT replSettings $
             -- MAYBE: use cutoff or partialIterT to prevent runaway
             D.iterTM handleDbg $ D.runDriver nterm
 
 -- only using one breakpoint at a time allowed at the moment
-type Breakpoints = SrcSpan
+newtype MaybeBreakpoint = MaybeBreakpoint { _fromMaybeBreakpoint :: Maybe SrcSpan }
 type DAnn = SrcSpan
-instance D.Breakpointable DAnn Breakpoints where
+instance D.Breakpointable DAnn MaybeBreakpoint where
     hasBreakpoints = error "Not implemented: Breakpointable DAnn Breakpoints"
 
 -- Peel off one layer
 handleDbg :: (Cek.PrettyUni uni fun, D.GivenCekReqs uni fun DAnn RealWorld)
-          => D.DebugF uni fun DAnn Breakpoints (Repl.InputT IO a)
-          -> Repl.InputT IO a
+          => D.DebugF uni fun DAnn MaybeBreakpoint (Repl.InputT IO ())
+          -> Repl.InputT IO ()
 handleDbg = \case
-    -- Note that we first turn Cek to IO and then `liftIO` it to InputT; the alternative of
-    -- directly using MonadTrans.lift needs MonadCatch+MonadMask instances for CekM, i.e. messy
-    D.StepF prevState k  -> liftIO (D.cekMToIO $ D.handleStep prevState) >>= k
+    D.StepF prevState k  -> do
+        -- Note that we first turn Cek to IO and then `liftIO` it to InputT; the alternative of
+        -- directly using MonadTrans.lift needs MonadCatch+MonadMask instances for CekM, i.e. messy
+        eNewState <- liftIO $ D.cekMToIO $ D.tryHandleStep prevState
+        case eNewState of
+            Right newState -> k newState
+            Left e         -> Repl.outputStrLn $ show e
```

Comment:
Would it be better to set the new state to `Terminating` (or a new state, say `Error`), and continue with `k`? In your implementation, once there's an error, the `driverThread` terminates and stops responding to user inputs, which is probably fine, but it's different from how terminating normally works (where it enters the `Terminating` state).

If anything, it should be the opposite - when terminating normally, the `driverThread` can terminate, but when there's an error, the user may want to do additional things like examing the stack frames etc.

Reply:
You are right, what happens when the error occurs is currently not very good. But I am not sure that if what you are proposing is the best approach. I will add a TODO note and we should discuss it further for another PR

Reply:
Yeah, sounds good.